### ISSUE TITLE
Bug: RunContinuationsAsynchronously does not run continuations asynchronously - more cases

### ISSUE THREAD

**Issue by sjb-sjb** [dotnet/coreclr#20087]
It [has been reported](https://support.microsoft.com/en-ca/help/3118695/runcontinuationsasynchronously-does-not-run-continuations-asynchronous) that RunContinuationsAsynchronously does not run continuations asynchronously when Task.WhenAll, Task.WhenAny, or TaskExtensions.Unwrap are used to create the continuations.  

The below shows that RunContinuationsAsynchronously also does not run continuations asynchronously when just await is used, and the TaskContinuationSource is constructed within a wrapper object. 

Steps to reproduce: using Visual Studio 15.8.4, create a blank UWP application with minimum and target versions set to 1803 (10.0; build 17134). Check that Microsoft.NETCore.UniversalWindowsPlatform version 6.1.7 is being used. Replace the MainPage.xaml.cs with the code below, and create two buttons in the MainPage.xaml that invoke the two button handlers. 

The code starts a task that waits on a TaskContinuationSource.Task, and another task that delays and then sets the result of the TaskContinuationTask. This is done three times (sections 1, 2, and 3 in the code). Indenting indicates which of the three sections is producing a given output line. In the second of the three sections, the TaskContinuationSource is constructed within a wrapper object rather than directly in the calling method. All the tasks are asynchronous. In one version the tasks run on the thread pool while in another version they run on the UI thread. 

Pressing the two buttons gives output like the following. This output is incorrect because of the places where "awaited on 6" or "awaited on 3" occurs before rather than after the corresponding "signaled (rc = True) on 6" or "signaled (rc = True) on 3" line. 

For the first button the output: 

<code>
    <pre>
   Testing on 3; use thread pool = True
   awaiting on 7
      awaiting on 5
         awaiting on 6
         signaling on 4
         awaited on 5
         signaled (rc = True) on 4
   signaling on 7
   signaled (rc = True) on 7
   awaited on 4
      signaling on 6
      awaited on 6
      signaled (rc = True)on 6
   </pre>
</code>

For the second button the output: 

<code>
   <pre>   
   Testing on 3; use thread pool = False
   awaiting on 3
      awaiting on 3
         awaiting on 3
         signaling on 3
         signaled (rc = True) on 3
   signaling on 3
   signaled (rc = True) on 3
      signaling on 3
      awaited on 3
      signaled (rc = True)on 3
         awaited on 3
   awaited on 3
   </pre>
</code>

I am not sure whether or not the use of a wrapper class is intrinsic to the bug. I thought I saw a rare case where the bug occurred in one of the non-wrapper sections; but I was then not able to reproduce it. A large number of trials would be required to test this further. 

There are also variants on section 2, in which the TaskCompletionSource is constructed in the calling method before the wrapper object is constructed. The behaviour of section 2 then seems to be similar to sections 1 and 3. 

This could be a compiler bug rather than a runtime bug. 

Here is the source code: 

    using System;
    using System.Collections.Generic;
    using System.Diagnostics;
    using System.IO;
    using System.Linq;
    using System.Runtime.InteropServices.WindowsRuntime;
    using System.Threading.Tasks;
    using Windows.Foundation;
    using Windows.Foundation.Collections;
    using Windows.UI.Xaml;
    using Windows.UI.Xaml.Controls;
    using Windows.UI.Xaml.Controls.Primitives;
    using Windows.UI.Xaml.Data;
    using Windows.UI.Xaml.Input;
    using Windows.UI.Xaml.Media;
    using Windows.UI.Xaml.Navigation;
    using System.Threading;
    using Windows.UI.Core;
    using Windows.ApplicationModel.Core;

    namespace App1
    {
        static public class ThreadHelper
        {
            static CoreDispatcher _dispatcher;

            static private async void FireAndForgetOnUiThread( this Func<Task> actionAsync)
            {
                _dispatcher = _dispatcher ?? CoreApplication.MainView.CoreWindow.Dispatcher;
                DispatchedHandler invokeAndHandle = async () => { try { await actionAsync(); } catch (Exception e) { Debug.WriteLine($"Exception on {Thread.CurrentThread.ManagedThreadId}: {e.Message}"); } };
                try {
                    await _dispatcher.RunAsync(CoreDispatcherPriority.Normal, invokeAndHandle).AsTask();
                } catch (Exception) {
                    Debug.Assert(false); // doesn't happen
                }
            }

            static private void FireAndForgetOnThreadPool( this Func<Task> actionAsync)
            {
                Action invokeAndHandle = async () => { try { await actionAsync(); } catch (Exception e) { Debug.WriteLine($"Exception on {Thread.CurrentThread.ManagedThreadId}: {e.Message}"); } };
                Task.Run(invokeAndHandle);
            }

            static public void FireAndForget( this Func<Task> actionAsync, bool onThreadPool)
            {
                if (onThreadPool) {
                    actionAsync.FireAndForgetOnThreadPool();
                } else {
                    actionAsync.FireAndForgetOnUiThread();
                }
            }

            static public void FireAndForget(this Action action, bool onThreadPool)
            {
                Func<Task> actionAsync = () => { action(); return Task.CompletedTask; };
                actionAsync.FireAndForget(onThreadPool); // not very efficient, but easy to write.
            }
        }

        public class TcsWrapper
        {
            public TcsWrapper()
            {
                this.TaskCompletionSource = new TaskCompletionSource<int>(TaskContinuationOptions.RunContinuationsAsynchronously);
            }
            public TcsWrapper(TaskCompletionSource<int> tcs)
            {
                this.TaskCompletionSource = tcs;
            }

            public TaskCompletionSource<int> TaskCompletionSource;
        }

        public sealed partial class MainPage : Page
        {
            public MainPage()
            {
                this.InitializeComponent();
            }

            private void Button_Click(object sender, RoutedEventArgs e) => this.TestTcs(true);

            private void Button_Click_1(object sender, RoutedEventArgs e) => this.TestTcs( false);

            private void TestTcs(bool onThreadPool)
            {
                Debug.WriteLine("");
                Debug.WriteLine($"Testing on {Thread.CurrentThread.ManagedThreadId}; use thread pool = {onThreadPool}");

                // *** Section 1 ***********************************************************
                Action tcsTest1 = () => {
                    TaskCompletionSource<int> tcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);

                    Func<Task> waitTcsAsync = async () => {
                        Debug.WriteLine($"awaiting on {Thread.CurrentThread.ManagedThreadId}");
                        await tcs.Task;
                        Debug.WriteLine($"awaited on {Thread.CurrentThread.ManagedThreadId}");
                    };
                    waitTcsAsync.FireAndForget(onThreadPool);

                    Func<Task> signalTcs = async () => {
                        await Task.Delay(500);
                        Debug.WriteLine($"signaling on {Thread.CurrentThread.ManagedThreadId}");
                        bool rc = tcs.TrySetResult(0);
                        Debug.WriteLine($"signaled (rc = {rc}) on {Thread.CurrentThread.ManagedThreadId}");
                    };
                    signalTcs.FireAndForget(onThreadPool);
                };
                tcsTest1.FireAndForget(onThreadPool);

                // *** Section 2 ***********************************************************
                Action tcsTest2 = () => {
                    TcsWrapper wrapper = new TcsWrapper();
                    //TcsWrapper wrapper = new TcsWrapper(new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously));
                    //TcsWrapper wrapper = new TcsWrapper() { TaskCompletionSource = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously) };

                    Func<Task> waitTcsAsync = async () => {
                        Debug.WriteLine($"   awaiting on {Thread.CurrentThread.ManagedThreadId}");
                        await wrapper.TaskCompletionSource.Task; // Was: await waitFlag.WaitAsync();
                        Debug.WriteLine($"   awaited on {Thread.CurrentThread.ManagedThreadId}");
                    };
                    waitTcsAsync.FireAndForget(onThreadPool);

                    Func<Task> signalTcs = async () => {
                        await Task.Delay(500);
                        Debug.WriteLine($"   signaling on {Thread.CurrentThread.ManagedThreadId}");
                        bool rc = wrapper.TaskCompletionSource.TrySetResult(0); // was: waitFlag.TrySet(); // was: waitFlag.Value = 1;
                        Debug.WriteLine($"   signaled (rc = {rc})on {Thread.CurrentThread.ManagedThreadId}");
                    };
                    signalTcs.FireAndForget(onThreadPool);
                };
                tcsTest2.FireAndForget(onThreadPool);

                // *** Section 3 ***********************************************************
                Action tcsTest3 = () => {
                    TaskCompletionSource<int> tcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);

                    Func<Task> waitTcsAsync = async () => {
                        Debug.WriteLine($"      awaiting on {Thread.CurrentThread.ManagedThreadId}");
                        await tcs.Task;
                        Debug.WriteLine($"      awaited on {Thread.CurrentThread.ManagedThreadId}");
                    };
                    waitTcsAsync.FireAndForget(onThreadPool);

                    Func<Task> signalTcs = async () => {
                        await Task.Delay(500);
                        Debug.WriteLine($"      signaling on {Thread.CurrentThread.ManagedThreadId}");
                        bool rc = tcs.TrySetResult(0);
                        Debug.WriteLine($"      signaled (rc = {rc}) on {Thread.CurrentThread.ManagedThreadId}");
                    };
                    signalTcs.FireAndForget(onThreadPool);
                };
                tcsTest3.FireAndForget(onThreadPool);

            }

        }
    }

**Comment by sjb-sjb** [comment]
Closing this issue as it appears to have the same cause as discussed in https://github.com/dotnet/corefx/issues/29664. 
Namely, passing TaskContinuationOptions.RunContinuationsAsynchronously into the constructor instead of TaskCreationOptions.RunContinuationsAsynchronously. 

You can see the bug in my test code above. 

Would suggest that the signature of the constructor be changed to prevent this kind of mistake.
